# Project 4 -  Predictive Analysis using scikit-learn

In [139]:
#importing the libraries
import numpy as np
import pandas as pd
import sklearn 
from pandas import DataFrame
from sklearn import preprocessing
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

In [129]:
#Reading the file
df_mushrooms = pd.read_csv("C:/Users/KATERINE/Documents/cuny classes/DATA_ACQ_MAN/week13/agaricus-lepiota.data")
df_mushrooms.head()

p  x  s  n  t p.1  f  c n.1  k  ... s.2  w w.1 p.2 w.2  o p.3 k.1 s.3  u
0  e  x  s  y  t   a  f  c   b  k  ...   s  w   w   p   w  o   p   n   n  g
1  e  b  s  w  t   l  f  c   b  n  ...   s  w   w   p   w  o   p   n   n  m
2  p  x  y  w  t   p  f  c   n  n  ...   s  w   w   p   w  o   p   k   s  u
3  e  x  s  g  f   n  f  w   b  k  ...   s  w   w   p   w  o   e   n   a  g
4  e  x  y  y  t   a  f  c   b  n  ...   s  w   w   p   w  o   p   k   n  g

[5 rows x 23 columns]

In [130]:
#Creating the columns for the dataframe.   Columns' names were taken from the "agaricus-lepiota.names" file
column_names = [
    'class', 'cap shape', 'cap surface', 'cap color', 'bruised', 'odor',
    'gill attachment', 'gill spacing', 'gill size', 'gill color', 
    'stalk shape', 'stalk_root', 'stalk surface above ring',
    'stalk surface below ring', 'stalk color above ring',
    'stalk color below ring', 'veil type', 'veil color', 'ring number',
    'ring type', 'spore print color', 'population', 'habitat'
]

df_mushrooms.columns = column_names

In [132]:
#Turning the columns into dummy variables
df_mushrooms_enc = pd.get_dummies(df_mushrooms)
df_mushrooms_enc

class_e  class_p  cap shape_b  cap shape_c  cap shape_f  cap shape_k  \
0           1        0            0            0            0            0   
1           1        0            1            0            0            0   
2           0        1            0            0            0            0   
3           1        0            0            0            0            0   
4           1        0            0            0            0            0   
5           1        0            1            0            0            0   
6           1        0            1            0            0            0   
7           0        1            0            0            0            0   
8           1        0            1            0            0            0   
9           1        0            0            0            0            0   
10          1        0            0            0            0            0   
11          1        0            1            0            0            0   
12          0        1            0            0            0            0   
13          1        0            0            0            0            0   
14          1        0            0            0            0            0   
15          1        0            0            0            1            0   
16          0        1            0            0            0            0   
17          0        1            0            0            0            0   
18          0        1            0            0            0            0   
19          1        0            1            0            0            0   
20          0        1            0            0            0            0   
21          1        0            1            0            0            0   
22          1        0            1            0            0            0   
23          1        0            1            0            0            0   
24          0        1            0            0            1            0   
25          1        0            0            0            0            0   
26          1        0            0            0            0            0   
27          1        0            0            0            1            0   
28          1        0            0            0            0            0   
29          1        0            1            0            0            0   
...       ...      ...          ...          ...          ...          ...   
8093        1        0            1            0            0            0   
8094        0        1            0            0            0            0   
8095        1        0            0            0            0            1   
8096        0        1            0            0            0            1   
8097        0        1            0            0            0            1   
8098        1        0            0            0            0            1   
8099        1        0            0            0            1            0   
8100        0        1            0            0            0            1   
8101        1        0            0            0            0            0   
8102        1        0            0            0            0            1   
8103        1        0            0            0            0            1   
8104        1        0            0            0            0            1   
8105        1        0            0            0            0            1   
8106        1        0            0            0            0            0   
8107        0        1            0            0            0            1   
8108        1        0            1            0            0            0   
8109        1        0            0            0            0            0   
8110        1        0            0            0            0            1   
8111        1        0            0            0            0            1   
8112        0        1      

# Predicting  whether or not a mushroom is poisonous based on odor and spore print color

### Naive Bayes (NB) Algorithm to evaluate the 'spore print color' column

In [133]:
#Organizing the data
label_names = ['Edible', 'Poison']
labels = df_mushrooms_enc['class_e']
feature_names=['Spore_Print_Color']
features=df_mushrooms_enc[['spore print color_k','spore print color_n','spore print color_b','spore print color_h',
    'spore print color_r','spore print color_o','spore print color_u','spore print color_w','spore print color_y']]

#splitting the data
train, test, train_labels, test_labels, = train_test_split(features, labels, test_size=0.33, random_state=42)


#initializing the clasiffier
gnb = GaussianNB()

#trainning the classifier
model = gnb.fit(train, train_labels)

#Making  predictions
preds=gnb.predict(test)
print(preds)

[0 0 0 ... 0 0 0]


In [134]:
#Evaluating accuracy
print(accuracy_score(test_labels, preds))


0.5042894442372249


### Naive Bayes (NB) Algorithm to evaluate the 'odor' column

In [135]:
#Organizing the data
label_names = ['Edible', 'Poison']
labels = df_mushrooms_enc['class_e']
feature_names=['Odor']
features=df_mushrooms_enc[['odor_a','odor_l','odor_c','odor_y','odor_f','odor_m','odor_n','odor_p',
    'odor_s']]

#splitting the data
train, test, train_labels, test_labels, = train_test_split(features, labels, test_size=0.33, random_state=42)


#initializing the clasiffier
gnb = GaussianNB()

#trainning the classifier
model = gnb.fit(train, train_labels)

#Making predictions
preds=gnb.predict(test)
print(preds)

[0 1 0 ... 0 1 0]


In [136]:
#Evaluating accuracy
print(accuracy_score(test_labels, preds))

0.9847071988064156


### Naive Bayes (NB) Algorithm to evaluate the 'spore print color' and 'odor' columns together

In [137]:
#Organizing the data
label_names = ['Edible', 'Poison']
labels = df_mushrooms_enc['class_e']
feature_names=['Spore_Print_Color', 'Odor']
features=df_mushrooms_enc[['spore print color_k','spore print color_n','spore print color_b','spore print color_h',
    'spore print color_r','spore print color_o','spore print color_u','spore print color_w','spore print color_y', 'odor_a','odor_l','odor_c','odor_y','odor_f','odor_m','odor_n','odor_p',
    'odor_s']]

#splitting the data
train, test, train_labels, test_labels, = train_test_split(features, labels, test_size=0.33, random_state=42)


#initializing the clasiffier
gnb = GaussianNB()

#trainning the classifier
model = gnb.fit(train, train_labels)

#Making predictions
preds=gnb.predict(test)
print(preds)

[0 1 0 ... 0 1 0]


In [138]:
#Evaluating accuracy
print(accuracy_score(test_labels, preds))

0.990302126072361


### Conclusion

* 'spore print color' and 'odor' columns together have 99% accuracy to predict whether or not a mushroom is poisonous.

* 'odor' column is 98% accurate to predict whether or not a mushroom is poisonous.

* 'spore print color' column is 50% accurate to predict whether or not a mushroom is poisonous.
